### PR TITLE
 Make Docker compose files to be able to use them directly in Portainer

### PR DESCRIPTION
**Pull request by sbrunner**
## Proposed change

This is an alternative of #2661, by don't have a separated set of docker compose files.

Note that's for discussion, and the documentation isn't updated.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [ ] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Review by stumpylog — CHANGES_REQUESTED** [review]
As mentioned, my biggest problem is the giant listing of all environment.  There's a few other specific things, but I don't see that listing as required by anything documented.

File: docker/compose/docker-compose.lib.yaml [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      - MARIADB_RANDOM_ROOT_PASSWORD
+      - MARIADB_ROOT_HOST
+      - MARIADB_MYSQL_LOCALHOST_USER
+      - MARIADB_MYSQL_LOCALHOST_GRANTS
+      - MARIADB_PASSWORD_HASH
+      - MARIADB_INITDB_SKIP_TZINFO
+      - MARIADB_AUTO_UPGRADE
+      - MARIADB_DISABLE_UPGRADE_BACKUP
+
+  paperless: &paperless
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/compose/docker-compose.lib.yaml]: No, removed, thanks :-)

File: docs/setup.md [hunk 2]
```diff
@@ -93,16 +94,14 @@ steps described in [Docker setup](#docker_hub) automatically.
...
     use from the default (8000):
 
-    ```yaml
-    ports:
-      - 8000:8000
+    ```
+    PAPERLESS_PORT=8000
     ```
 
     Replace the part BEFORE the colon with a port of your choice:
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docs/setup.md:100]: +1 :-)

**Comment by shamoon** [comment]
Yea agree about the massive environment list, this version also uses the extends feature which [as I mentioned in the other PR](https://github.com/paperless-ngx/paperless-ngx/pull/2660#pullrequestreview-1294759720) I am a little concerned will confuse folks

Feels like competing needs and goals here in terms of codebase maintenance / user ease-of-use / better comparability with portainer. but I’m not sure which things wins out in the end…

File: docker/compose/docker-compose.lib.yaml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    restart: unless-stopped
+    volumes:
+      - ${REDIS_DATA_VOLUME}:/data
+
+  postgres:
+    image: docker.io/library/postgres:13
+    restart: unless-stopped
+    volumes:
+      - ${POSGRES_DATA_VOLUME}:/var/lib/postgresql/data
+    environment:
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/compose/docker-compose.lib.yaml:17]: The list is relay less verbose when we have some pass throw environment variable:
e.g.:
`- PAPERLESS_OCR_LANGUAGE`
become
`PAPERLESS_OCR_LANGUAGE: ${PAPERLESS_OCR_LANGUAGE}`
I don't think that's better...

File: docker/compose/docker-compose.lib.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    image: docker.io/library/mariadb:10
+    restart: unless-stopped
+    volumes:
+      - ${MARIADB_DATA_VOLUME}:/var/lib/mysql
+    environment:
+      - MARIADB_DATABASE=paperless
+      - MARIADB_USER=paperless
+      - MARIADB_PASSWORD=paperless
+      - MARIADB_ROOT_PASSWORD=paperless
+      - MARIADB_ROOT_PASSWORD_HASH
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/compose/docker-compose.lib.yaml:41]: > It's a massive amount of work to maintain this list for every image

I think that's largely exaggerate...

> I don't see anything that says it is required to list everything.

It's pass throw of variable that can be defined in the .env file then yes it's required

> It is intimidating to a user to see

The user doesn't have to open this file, then it's largely more a simplification for them than a complexity

And we should keep in mind that to largely simplify the installation on a NAS, where currently many users abandoned to do that.

> It includes many options most users never need to set or know about

Effectively, I put all the option, but we should keep in mind that the option is missing the user can't set it

File: docker/compose/docker-compose.lib.yaml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      - MARIADB_PASSWORD_HASH
+      - MARIADB_INITDB_SKIP_TZINFO
+      - MARIADB_AUTO_UPGRADE
+      - MARIADB_DISABLE_UPGRADE_BACKUP
+
+  paperless: &paperless
+    image: ghcr.io/paperless-ngx/paperless-ngx:latest
+    restart: unless-stopped
+    ports:
+      - ${PAPERLESS_PORT}:${PAPERLESS_INNER_PORT}
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/compose/docker-compose.lib.yaml:56]: It's not what I see, when I change the port, I should also change the PAPERLESS_URL and that change the inner port...
It's possible that's better to do differently, but first we should decide if we want this pull request or not...

**Comment by stumpylog** [comment]
We should definitely fix whatever is wrong with Portainer's specific file and the specific steps for setting it up also seem useful.  

That said, I'd prefer to have the files be explain with what they support and the environment they configure.  It's duplicated code, yes, but simple to update image version when wanted

File: docker/compose/docker-compose.lib.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (edenhaus) [reply docker/compose/docker-compose.lib.yaml:41]: > The user doesn't have to open this file, then it's largely more a simplification for them than a complexity
> 
> And we should keep in mind that to largely simplify the installation on a NAS, where currently many users abandoned to do that.

That's only true for people using portainer and/or trust files without even looking at it. Good practise is to check a file before using it.
Imo a docker-compose with only the few required option set is better than all options as users can get confused and lookup all options, which are not required at the beginning. I share the opinion of @shamoon

Is portainer supporting the docker-compose option `env_file` or similar? So use save the env variable of your stack in that specific file

**Comment by shamoon** [comment]
Of them all I think #2661 is the least "intrusive", it allows portainer users to use it very easily but keeps the portainer files separate and has the least amount of maintenance by using the `extends` feature (without changing the 'core' compose files).

It does still have the issue of the full ENV list which I agree is a pain to maintain. sbrunner you said that list is required for portainer compatibility but isn't it true that it would still work as long as you include the required ones and then users can specify any others at the level of each service (either in the compose file if they cloned it or just via portainer)? I tested with #2661 and it did go smoothly, afterwards I could sill go to the stack containers and add ENV variables. So in other words it would be slightly less convenient for the end-user but avoids us having to use an exhaustive list?

That would be my suggestion, essentially #2661 with a pared-down list

**Comment by sbrunner** [comment]
>  afterwards I could sill go to the stack containers and add ENV variables.

I try it => Using the button `Dupplicate/Edit` I can edit the env variable, `Deploy the container`, it looks good but when I enter the container, the added env variable is not here :-(

It's a shame because it would be an excellent solution...

**Comment by shamoon** [comment]
Would like to circle back to this and see if we can agree on a solution as these three PRs seem to be languishing here.

**Comment by shamoon** [comment]
@sbrunner Thank you for all of the effort. We talked about this a bit and as a group weren't quite convinced that the pros outweigh the cons with these changes. A few of us (myself included) are already using portainer with paperless successfully. Perhaps some of the info here could be added to the docs or linked to instead.

Of course if there is significant interest from the larger community let's re-evaluate but for now we'll close these PRs.

Welcome your feedback, thank you again for the work.

**Comment by sbrunner** [comment]
OK, thanks for your attention, it's not a problem for me, then I will create a separate repository for that, it's probably not a gad thing because it's more agile for deeper investigations :-)

**Pull request closed without merging**